server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -257,7 +257,9 @@ private static FieldsVisitor createStoredFieldsVisitor(
             boolean loadSource = sourceRequired(context);
             if (loadSource) {
                 if (false == sourceLoader.requiredStoredFields().isEmpty()) {
-                    return new CustomFieldsVisitor(sourceLoader.requiredStoredFields(), true);
+                    // add the stored fields needed to load the source mapping to an empty set so they aren't returned
+                    sourceLoader.requiredStoredFields().forEach(fieldName -> storedToRequestedFields.putIfAbsent(fieldName, Set.of()));
+                    return new CustomFieldsVisitor(storedToRequestedFields.keySet(), true);
```

Comment:
This change fixes that leaking `fields` in the search response. It's tricky!

Reply:
Sneaky.  Nice catch.